actors/evm/src/interpreter/precompiles/bls_util.rs

```diff
@@ -0,0 +1,421 @@
+    // SAFETY: all inputs are valid blst types
+    unsafe { blst_p2_mult(&mut result, &p_jacobian, scalar.b.as_ptr(), scalar.b.len() * 8) };
+
+    // Convert result back to affine coordinates
+    p2_to_affine(&result)
+}
+
+/// Encodes a G2 point in affine format into byte slice with padded elements.
+/// G2 points have two coordinates (x,y) where each coordinate is a complex number (real,imaginary)
+/// So we need to encode 4 field elements total: x.re, x.im, y.re, y.im
+pub(super) fn encode_g2_point(input: &blst_p2_affine) -> Vec<u8> {
+    // Create output buffer with space for all coordinates (4 * 64 bytes)
+    let mut out = vec![0u8; PADDED_G2_LENGTH];
+    // Encode x coordinate
+    // Real part (x.fp[0])
+    fp_to_bytes(&mut out[..PADDED_FP_LENGTH], &input.x.fp[0]);
+    // Imaginary part (x.fp[1])
+    fp_to_bytes(&mut out[PADDED_FP_LENGTH..2 * PADDED_FP_LENGTH], &input.x.fp[1]);
+
+    // Encode y coordinate
+    // Real part (y.fp[0])
+    fp_to_bytes(&mut out[2 * PADDED_FP_LENGTH..3 * PADDED_FP_LENGTH], &input.y.fp[0]);
+    // Imaginary part (y.fp[1])
+    fp_to_bytes(&mut out[3 * PADDED_FP_LENGTH..4 * PADDED_FP_LENGTH], &input.y.fp[1]);
+
+    out
+}
+
+/// Extracts a G2 point in Affine format from a 256 byte slice representation.
+///
+/// **Note**: This function will perform a G2 subgroup check if `subgroup_check` is set to `true`.
+///
+/// Subgroup checks are required for:
+/// - Scalar multiplication
+/// - Multi-scalar multiplication (MSM)
+/// - Pairing operations
+///
+/// But not required for:
+/// - Point addition
+/// - Point negation
+pub(super) fn extract_g2_input(
+    input: &[u8],
+    subgroup_check: bool,
+) -> Result<blst_p2_affine, PrecompileError> {
+    // Check input length (256 bytes = 4 * 64 bytes for x.re, x.im, y.re, y.im)
+    if input.len() != G2_INPUT_ITEM_LENGTH {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    let [x_re, x_im, y_re, y_im] = remove_g2_padding(input)?;
+    // Convert bytes to point
+    let point = decode_g2_on_curve(x_re, x_im, y_re, y_im)?;
+
+    if subgroup_check {
+        // Subgroup check (more expensive but required for certain operations)
+        // Verifies that the point has the correct order and is in G2
+        // SAFETY: point is properly initialized above
+        unsafe {
+            if !blst_p2_affine_in_g2(&point) {
+                return Err(PrecompileError::InvalidInput);
+            }
+        }
+    } else {
+        // Basic curve check (less expensive, sufficient for addition)
+        // Only verifies that the point is on the curve
+        // SAFETY: point is properly initialized above
+        if unsafe { !blst_p2_affine_on_curve(&point) } {
+            return Err(PrecompileError::InvalidInput);
+        }
+    }
+
+    Ok(point)
+}
+
+/// https://eips.ethereum.org/EIPS/eip-2537
+/// Encodes a single finite field element into byte slice with padding.
+pub(super) fn fp_to_bytes(out: &mut [u8], input: *const blst_fp) {
```

Comment:
This should me marked unsafe.

Reply:
Woah why are we not matching revm and inputting a val of type &blst_fp ?